lib/templates/lab2/lab2.dart

```diff
@@ -1,19 +1,69 @@
-class Book {
-  // your code here
+class Book with JsonSerializable {
+  final String title, author;
+  final int numberOfPages;
+  Book(this.title, this.author, this.numberOfPages);
+
+  String bookDetails() {
+    return 'Book "$title" by $author, number of pages: $numberOfPages';
+  }
 }
 
-class FictionBook {
-  // your code here
+class FictionBook extends Book {
+  String genre;
+  FictionBook(super.title, super.author, super.numberOfPages, this.genre);
+
+  @override
+  String bookDetails() {
+    return 'Book "$title" by $author, genre: $genre, number of pages: $numberOfPages';
+  }
 }
 
 mixin JsonSerializable {
-  // your code here
+  String? title, author;
+  int? numberOfPages;
+  Map toJson() {
+    return {title: this, author: this, numberOfPages: this};
+  }
 }
 
-double calculateArea(){
-  // your code here
+double calculateArea({double? length, double? width, String shape = "rectangle"}) {
+  double result = 0;
+  if (width != null && length != null) {
+    if (shape == "rectangle") {
+      result = width * length;
+    } else if (shape == "circle") {
+      result = 3.14 * length * length;
+    }
+  }
+  return result;
 }
 
 class User{
-  // your code here
+  String name;
+  int? age;
+  String? email;
+  User(this.name, this.age, this.email);
+
+  void printUserDetails() {
+    print("Name: $name Age: ${age ?? "N/A"} Email: ${email ?? "N/A"}");
+  }
+}
+
+void main() {
+  // Task 1
+  var book = Book("Title", "Author", 100);
+  print(book.bookDetails());
+  // Task 2
+  var fictionBook = FictionBook("Fiction book", "Author", 1000, "fantasy");
+  print(book.bookDetails());
+  // Task 3
+  print(book.toJson());
```

Comment:
Incorrect output, I should use 'this.title', not just 'this'
<img width="584" alt="Screenshot 2024-06-11 at 07 01 34" src="https://github.com/timur-harin/Sum24FlutterCourse/assets/64702874/c1b530dc-5880-40e3-9cbb-56c55ef80506">